### PR TITLE
ref(Datasets): List of tuples

### LINKED ISSUE #218: Change dataset partition to tuples of data points
Currently each dataset partition is defined as `tuple[A, b]`  where A is an `n x m` and b is an `n x 1` matrix where `n` is the number of datapoints and `m` is the feature dimension. Issues occur when the dataset is too large to fit into memory and we must lazily load it or when the datapoints are matricies, since we can't create these matricies. Therefore I suggest changing from this matrix representation to lists of tuples of datapoints `(x, y)`. 

This would require minor code changes, such as changing the cost functions which expect these matricies to stack the datapoints themselves and converting the already implemented datasets.

### PR DESCRIPTION
**Pull request by Simpag**
This PR converts the `DatasetPartitions` from two matrices of features and targets to a list of tuples where each tuple is a `(feature, target)`. This is required as some datasets such as images might not make sense to store as a matrix (if they are RGB images they consist of three matrices) or the dataset might be too large to fit into memory and thus is required to be lazily-loaded. 

This PR also includes two new datasets. `PyTorchWrapper` takes a PyTorch dataset and efficiently splits this dataset into multiple training partitions with a few options such as heterogeneity. `KaggleDataset` enables users to download and split datasets from [Kaggle](https://www.kaggle.com/). 

Lastly, this PR adds an optional `test_dataset` into the benchmark problem which is to be used by future evaluation metrics. 

Closes #218

### CODE REVIEW

File: decent_bench/costs/_base/_quadratic_cost.py [hunk 1]
```diff
@@ -20,18 +21,37 @@ class QuadraticCost(Cost):
...
 
-    def __init__(self, A: Array, b: Array, c: float):  # noqa: N803
-        self.A: NDArray[float64] = iop.to_numpy(A)
-        self.b: NDArray[float64] = iop.to_numpy(b)
+    def __init__(
+        self,
+        c: float,
+        A: Array | None = None,  # noqa: N803
+        b: Array | None = None,
+        dataset: DatasetPartition | None = None,
```
[the remark this answers is not in the document]
Comment (nicola-bastianello) [reply decent_bench/costs/_base/_quadratic_cost.py]: there will also be other costs that need data in different formats, which do not fit with dataset

File: decent_bench/datasets/_dataset.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+DatasetPartition: TypeAlias = list[Datapoint]  # noqa: UP040
+"""List of datapoints representing one dataset partition."""
+
+
+class Dataset(ABC):
+    """Dataset containing partitions in the form of lists of tuples (feature, target)."""
+
+    @abstractmethod
+    def training_partitions(self) -> Sequence[DatasetPartition]:
+        """Partitions used for finding the optimal optimization variable x."""
```
[the remark this answers is not in the document]
Comment (nicola-bastianello) [reply decent_bench/datasets/_dataset.py]: yes, great idea, let's have "get all data"

**Comment by nicola-bastianello** [comment]
> Total datapoints might be a bit confusing as we have training partitions and usually define the number of datapoints per partition. I was thinking about adding a test_partition function but that would cause trouble when you use like mnist which has a train and a test set (and many others). What could be of use would be shape of features and targets.

I was thinking it would be total number of datapoints before they get partitioned; in my mind, Dataset is a single block of data (and we have info like the total number + the shape of features and targets - nice idea!); then, as a functionality of the Dataset we provide `training_partitions` to split it. But once paritioning has happened, that's the output of the function, it's not a modification made on the original block of data. Does this make sense to you?

File: decent_bench/costs/_empirical_risk/_linear_regression_cost.py [hunk 3]
```diff
@@ -42,41 +42,41 @@ class LinearRegressionCost(EmpiricalRiskCost):
...
         """
-        if len(iop.shape(A)) != 2:
-            raise ValueError("Matrix A must be 2D")
-        if len(iop.shape(b)) != 1:
-            raise ValueError("Vector b must be 1D")
-        if iop.shape(A)[0] != iop.shape(b)[0]:
-            raise ValueError(f"Dimension mismatch: A has {iop.shape(A)[0]} rows but b has {iop.shape(b)[0]} elements")
-        if isinstance(batch_size, int) and (batch_size <= 0 or batch_size > iop.shape(A)[0]):
+        if len(iop.shape(dataset[0][0])) != 1:
+            raise ValueError(f"Dataset features must be vectors, got: {dataset[0][0]}")
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_empirical_risk/_linear_regression_cost.py]: I just realised that my previous comment was poorly formulated. What I meant to say is that most frameworks expect that the input does not contain any extra unused dimensions and throws an error if there are any. 

Either way, I can test what happens but I am pretty sure it will error out as the length of the shape tuple would be 2 in your example case. Another issue I just thought of is that the dataset might be infinite, it could be from some type of generator and we would not be able to squeeze the dimensions. Therefore I don't think we should perform any operations on the dataset

File: decent_bench/costs/_base/_quadratic_cost.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_base/_quadratic_cost.py]: Do you have any (simple) examples? The idea behind the dataset class is that every dataset in decent bench should be of the defined shape so that we can use any dataset with any cost function (to some degree, wont work in some cases. For example "infinite" datasets but those will most likely not be in decent bench and users able to create those should be able to understand where  and where not they can use them). We want some kind of known shape. The `Array` class is very loosely defined so it could be many different things.

File: decent_bench/costs/_empirical_risk/_linear_regression_cost.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicola-bastianello) [reply decent_bench/costs/_empirical_risk/_linear_regression_cost.py]: right, indeed the shape will have 2 elements and we get an error. then I'm ok with not doing anything, but let's document thoroughly what we expect the features and targets to look like in terms of shape

honestly I wouldn't worry for infinite datasets, I doubt they would ever be used in the context of decent-bench. but I agree that it might mess things up in the future if we do "silent" operations on the dataset

File: decent_bench/costs/_base/_quadratic_cost.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/costs/_base/_quadratic_cost.py]: I see what you mean. I wonder if we should create a new subclass or something similar for regularizers and try to keep the cost function class more based around the dataset structure? I think it would be a lot easier to use the framework if we try to keep the cost functions to use the same structures. However, I am not sure if we can make this work if there are more costs which require special shapes. Do you have any ideas or should we just go for the more loosely structured cost functions?

[the remark this answers is not in the document]
Comment (nicola-bastianello) [reply decent_bench/costs/_base/_quadratic_cost.py]: unfortunately I don't see a way to divide all possible costs in neat categories like this; e.g. see the cost we use [here](https://arxiv.org/pdf/2009.08335); it is an academic example, but still something we want to leave the freedom to do.

what we could do is: Cost remains the generic interface for anything that users might want to implement (and as you say, any example we implement is well documented to show how the "data" defining the cost is passed to the init); EmpiricalRiskCost instead is the subclass that deals with data-driven costs, and we require that all of them get a Dataset as input. this way Dataset becomes linked to EmpiricalRiskCost only. this could help clarifying the purpose of Dataset

one thing that just came to my mind: in some cases we deal with unsupervised learning, where the dataset is only a collection of features (e.g. when we do k-means clustering). I think it would be good to think about this now, instead of having to go back and make structural changes. so the question is: any idea on how we could allow Dataset to be unsupervised (that is, features only), while retaining the same interface? if there's no way we can create a separate class, but it would be nice to have only one

File: decent_bench/datasets/_dataset.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+
+from decent_bench.utils.types import DatasetPartition
+
+
+class Dataset(ABC):
+    """
+    Abstract wrapper for datasets used in distributed optimization benchmark problems.
```
Comment (nicola-bastianello) [inline decent_bench/datasets/_dataset.py]: distributed -> decentralized, also below

File: decent_bench/utils/types.py [hunk 5]
```diff
@@ -36,6 +38,12 @@
 or an integer specifying a single datapoint.
 """
 
+type Datapoint = tuple["Array", "Array"]  # noqa: TC008
+"""Tuple of (x, y) representing one datapoint where x are features and y is the target."""
+
+type DatasetPartition = list[Datapoint]
```
[the remark this answers is not in the document]
Comment (Simpag) [reply decent_bench/utils/types.py]: That sounds like a good middle ground for now. We can always refactor it in the future but as you said, we have some other more important features to complete.
